sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqUnboundedReader.java

```diff
@@ -0,0 +1,230 @@
+    this.connectionHandler = this.source.spec.connectionHandlerProviderFn().apply(null);
+    this.averageRecordSize = new MovingAvg();
+
+    RabbitMqCheckpointMark cpMark = checkpointMark;
+    if (cpMark == null) {
+      cpMark = new RabbitMqCheckpointMark(this.connectionHandler);
+    } else {
+      cpMark.setChannelLeaser(this.connectionHandler);
+    }
+    this.checkpointMark = cpMark;
+    ensureTimestampPolicySet(
+        Optional.ofNullable(checkpointMark).map(RabbitMqCheckpointMark::getWatermark));
+  }
+
+  private void ensureTimestampPolicySet(Optional<Instant> prevWatermark) {
+    if (this.timestampPolicy == null) {
+      this.timestampPolicy =
+          source.spec.timestampPolicyFactory().createTimestampPolicy(prevWatermark);
+    }
+  }
+
+  @Override
+  public Instant getWatermark() {
+    Instant watermark = timestampPolicy.getWatermark(context);
+    checkpointMark.setWatermark(watermark);
+    return watermark;
+  }
+
+  @Override
+  public UnboundedSource.CheckpointMark getCheckpointMark() {
+    return checkpointMark;
+  }
+
+  @Override
+  public RabbitMqSource getCurrentSource() {
+    return source;
+  }
+
+  @Override
+  public byte[] getCurrentRecordId() {
+    return Optional.ofNullable(currentRecord).map(recordIdPolicy::apply).get();
+  }
+
+  @Override
+  public Instant getCurrentTimestamp() {
+    return Optional.ofNullable(currentRecord).map(timestampPolicy::getTimestampForRecord).get();
+  }
+
+  @Override
+  public RabbitMqMessage getCurrent() {
+    if (currentRecord == null) {
+      throw new NoSuchElementException();
+    }
+    return currentRecord;
+  }
+
+  @Override
+  public boolean start() throws IOException {
+    try {
+      checkpointMark.startReading();
+
+      ChannelLeaser.UseChannelFunction<String> setupFn =
+          channel -> {
+            ReadParadigm paradigm = source.spec.readParadigm();
+            String queueName = paradigm.queueName();
+            if (paradigm instanceof ReadParadigm.NewQueue) {
+              ReadParadigm.NewQueue newQueue = (ReadParadigm.NewQueue) paradigm;
+
+              final boolean durable = true;
+              final boolean autoDelete = false;
+              final boolean exclusive = false;
+              final Map<String, Object> arguments = Collections.emptyMap();
+
+              // this may fail if the queue isn't actually new and has different properties
+              channel.queueDeclare(queueName, durable, exclusive, autoDelete, arguments);
+
+              // the default exchange does not allow for explicitly binding a queue
+              if (!"".equalsIgnoreCase(newQueue.getExchange())) {
+                channel.queueBind(queueName, newQueue.getExchange(), newQueue.getRoutingKey());
+              }
+            }
+
+            return queueName;
+          };
+
+      queueName = connectionHandler.useChannel(checkpointMark.getCheckpointId(), setupFn);
+    } catch (IOException e) {
+      checkpointMark.stopReading();
```

Comment:
Maybe, but I don't think so. I don't know all the intended semantics of `close` for a Reader in Beam, but I do know that I want to inform the checkpoint mark that nothing is attempting to read from it. It's possible Beam could call `start` again in which case this would definitely not be "closed".